### PR TITLE
Issue 15727 dynamically generate graphql schema

### PR DESCRIPTION
**Pull request by dsilvam**

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java [hunk 1]
```diff
@@ -364,67 +376,71 @@ public void delete(final Field field) throws DotDataException {
...
+      permissionAPI.checkPermission(type, PermissionLevel.EDIT_PERMISSIONS, user);
 
-		  Structure structure = new StructureTransformer(type).asStructure();
-		  com.dotmarketing.portlets.structure.model.Field legacyField = new LegacyFieldTransformer(field).asOldField();
+      Field oldField = fieldFactory.byId(field.id());
+      if(oldField.fixed() || oldField.readOnly()){
+          throw new DotDataException("You cannot delete a fixed or read only field");
+      }
 
+      Structure structure = new StructureTransformer(type).asStructure();
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java]: Make this variable final

File: dotCMS/src/main/java/com/dotcms/graphql/CategoryToMapTransformer.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        }
+        try {
+            this.cats = APILocator.getCategoryAPI().getParents(con, user, true);
+        } catch (DotDataException | DotSecurityException e) {
+            throw new DotStateException(e);
+        }
+        final Map<String, Object> newMap = new HashMap<>();
+        if(cats!=null && !cats.isEmpty()) {
+
+            for (final Field field : con.getContentType().fields()) {
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/graphql/CategoryToMapTransformer.java]: here you can use a lambda expression

File: dotCMS/src/main/java/com/dotcms/graphql/CategoryToMapTransformer.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        if (con.getInode() == null) {
+            throw new DotStateException("Contentlet needs an inode to get fields");
+        }
+        try {
+            this.cats = APILocator.getCategoryAPI().getParents(con, user, true);
+        } catch (DotDataException | DotSecurityException e) {
+            throw new DotStateException(e);
+        }
+        final Map<String, Object> newMap = new HashMap<>();
+        if(cats!=null && !cats.isEmpty()) {
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/graphql/CategoryToMapTransformer.java]: this line can be replaced by `if (UtilMethods.isSet(cats))`

File: dotCMS/src/main/java/com/dotcms/graphql/CategoryToMapTransformer.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+        final Map<String, Object> map = new HashMap<>();
+
+        try {
+
+            final Category parentCategory        = APILocator.getCategoryAPI().find(field.values(), user, true);
+            final List<Map<String, Object>> childCategories = new ArrayList<>();
+
+            if(parentCategory != null) {
+                for (Category category : cats) {
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/graphql/CategoryToMapTransformer.java]: this logic can be replaced by a lambda expression

File: dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java [hunk 5]
```diff
@@ -364,67 +376,71 @@ public void delete(final Field field) throws DotDataException {
...
-	      ) {
-	    	  this.contentletAPI.cleanField(structure, legacyField, this.userAPI.getSystemUser(), false);
-	      }
+      fieldFactory.moveSortOrderBackward(type.id(), oldField.sortOrder());
+      fieldFactory.delete(field);
 
-          fieldFactory.moveSortOrderBackward(type.id(), oldField.sortOrder());
-          fieldFactory.delete(field);
+      ActivityLogger.logInfo(ActivityLogger.class, "Delete Field Action",
+          String.format("User %s/%s eleted field %s from %s Content Type.", user.getUserId(), user.getFirstName(),
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java]: This message should be fixed

File: dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java (hunk 1 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java]: old code but done

File: dotCMS/src/main/java/com/dotcms/graphql/CategoryToMapTransformer.java (hunk 2 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/graphql/CategoryToMapTransformer.java]: I borrow all this code (entire CategoryToMapTransformer) from Fabrazzio's branch. Won't touch it for now until merged.

File: dotCMS/src/main/java/com/dotcms/graphql/CategoryToMapTransformer.java (hunk 3 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/graphql/CategoryToMapTransformer.java]: same as above

File: dotCMS/src/main/java/com/dotcms/graphql/CategoryToMapTransformer.java (hunk 4 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/graphql/CategoryToMapTransformer.java]: same as above

File: dotCMS/src/main/java/com/dotcms/graphql/business/GraphqlAPIImpl.java [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+
+    @Override
+    public void invalidateSchema() {
+        this.schema = null;
+    }
+
+    private void printSchema() {
+        SchemaPrinter printer = new SchemaPrinter();
+        try {
+            Files.write(Paths.get("/Users/danielsilva/Documents/schema.graphqls"), printer.print(schema).getBytes());
```
[the remark this answers is not in the document]
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/graphql/business/GraphqlAPIImpl.java]: fixed

File: dotCMS/src/main/java/com/dotcms/contenttype/business/ContentTypeAPIImpl.java [hunk 7]
```diff
@@ -41,25 +45,27 @@
   private final User user;
   private final Boolean respectFrontendRoles;
   private final FieldAPI fieldAPI;
+  private final LocalSystemEventsAPI localSystemEventsAPI;
 
 
 
   public ContentTypeAPIImpl(User user, boolean respectFrontendRoles, ContentTypeFactory fac, FieldFactory ffac,
-      PermissionAPI perms, FieldAPI fAPI) {
+      PermissionAPI perms, FieldAPI fAPI, final LocalSystemEventsAPI localSystemEventsAPI) {
```
[the remark this answers is not in the document]
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/contenttype/business/ContentTypeAPIImpl.java]: yeah already covered

File: dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java (hunk 5 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java]: fixed

**Review by nollymar — APPROVED** [review]

**Review by erickgonzalez — APPROVED** [review]

**Pull request merged**